backend/windmill-api/src/database_triggers/trigger.rs

```diff
@@ -0,0 +1,474 @@
+            options
+        );
+
+        Ok((
+            self.0
+                .copy_both_simple::<bytes::Bytes>(query.as_str())
+                .await
+                .map_err(Error::Postgres)?,
+            LogicalReplicationSettings::new(false),
+        ))
+    }
+
+    async fn send_status_update(
+        primary_keep_alive: PrimaryKeepAliveBody,
+        copy_both_stream: &mut Pin<&mut CopyBothDuplex<Bytes>>,
+    ) {
+        let mut buf = BytesMut::new();
+        let ts = chrono::Utc.with_ymd_and_hms(2000, 1, 1, 0, 0, 0).unwrap();
+        let ts = chrono::Utc::now()
+            .signed_duration_since(ts)
+            .num_microseconds()
+            .unwrap_or(0);
+
+        buf.put_u8(b'r');
+        buf.put_u64(primary_keep_alive.wal_end);
+        buf.put_u64(primary_keep_alive.wal_end);
+        buf.put_u64(primary_keep_alive.wal_end);
+        buf.put_i64(ts);
+        buf.put_u8(0);
+        copy_both_stream.send(buf.freeze()).await.unwrap();
+        tracing::debug!("Send update status message");
+    }
+}
+
+async fn update_ping(
+    db: &DB,
+    database_trigger: &DatabaseTrigger,
+    error: Option<&str>,
+) -> Option<()> {
+    match sqlx::query_scalar!(
+        r#"
+        UPDATE 
+            database_trigger
+        SET 
+            last_server_ping = now(),
+            error = $1
+        WHERE
+            workspace_id = $2
+            AND path = $3
+            AND server_id = $4 
+            AND enabled IS TRUE
+        RETURNING 1
+        "#,
+        error,
+        &database_trigger.workspace_id,
+        &database_trigger.path,
+        *INSTANCE_NAME
+    )
+    .fetch_optional(db)
+    .await
+    {
+        Ok(updated) => {
+            if updated.flatten().is_none() {
+                // allow faster restart of database trigger
+                sqlx::query!(
+                    r#"
+                UPDATE 
+                    database_trigger 
+                SET
+                    last_server_ping = NULL 
+                WHERE 
+                    workspace_id = $1 
+                    AND path = $2 
+                    AND server_id IS NULL"#,
+                    &database_trigger.workspace_id,
+                    &database_trigger.path,
+                )
+                .execute(db)
+                .await
+                .ok();
+                tracing::info!(
+                    "Database {} changed, disabled, or deleted, stopping...",
+                    database_trigger.path
+                );
+                return None;
+            }
+        }
+        Err(err) => {
+            tracing::warn!(
+                "Error updating ping of database {}: {:?}",
+                database_trigger.path,
+                err
+            );
+            return None;
+        }
+    };
+
+    Some(())
+}
+
+async fn loop_ping(db: &DB, database_trigger: &DatabaseTrigger, error: Option<&str>) {
+    loop {
+        if update_ping(db, database_trigger, error).await.is_none() {
```

Comment:
I would dispatch the error through `update_ping` and move the `warn!` here instead. So it's easier to understand why the loop is stopped